### PR TITLE
Implemented Layout/SpaceInsideParens with EnforcedStyle: compact

### PR DESCRIPTION
**Pull request by itay-grudev**
```yml
Layout/SpaceInsideParens:
  EnforcedStyle: compact
```

```ruby
@example 
# The `compact` style enforces that parentheses have a space at the
# beginning with the exception that successive right parentheses are allowed.
# Note: Empty parentheses should not have spaces.

# bad
f(3)
g = (a + 3)
y( )
g( f( x ) )
g( f( x( 3 ) ), 5 )

# good
f( 3 )
g = ( a + 3 )
y()
g( f( x ))
g( f( x( 3 )), 5 )
```

-----------------

Before submitting the PR make sure the following are checked:

* [X] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [X] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/layout/space_inside_parens.rb
Comment (dvandersluis) [inline lib/rubocop/cop/layout/space_inside_parens.rb:122]:
```suggestion
        def each_extraneous_space_between_right_parents(token1, token2)
```

Or just use `extra`?

File: lib/rubocop/cop/layout/space_inside_parens.rb
Comment (dvandersluis) [inline lib/rubocop/cop/layout/space_inside_parens.rb:125]: You should extract the `range_between` call into a variable rather than call it twice.

File: changelog/new_space_inside_parens_compact.md
Comment (dvandersluis) [inline changelog/new_space_inside_parens_compact.md:1]: This needs to link to this PR.
```suggestion
* [#10025](https://github.com/rubocop/rubocop/pull/10025) Changed cop `SpaceInsideParens` to include a `compact` style.  ([@itay-grudev][])
```

**Comment by dvandersluis** [comment]
I'm not sure I understand why it'd be desired to compact only right parens? If this is a style that people use, should it not apply on both sides? Or else, the style should probably be called `compact_right` or something similar. I'm not sure if "compact" even makes sense here when most of the parentheses have white space around them, which makes them not compact 😄 Maybe `collapse`/`collapse_right`?

Also you'd need to update `VersionChanged` in default.yml.

File: config/default.yml
Comment (dvandersluis) [inline config/default.yml:1350]:
```suggestion
  VersionChanged: '<<next>>'
```

Sorry should have specified this.

Comment (itay-grudev) [reply config/default.yml]: Out of curiosity, how is this replaced afterwards? Job that runs on pull_requests?

Comment (dvandersluis) [reply config/default.yml]: There's a `cut_release` task that replaces all the `<<next>>`s with whatever the next version ends up being.

**Comment by dvandersluis** [comment]
I'm not convinced that this is a style that makes sense for RuboCop as a whole. You could implement it as a custom cop for your codebase/company and just disable `Layout/SpaceInsideParens`.

@bbatsov what do you think?

**Comment by itay-grudev** [comment]
RubCop is a collection of rules, configurable to suit different projects, different teams with different views. The objective of the project is internal consistency, with a healthy set of defaults. If there was one style that was suitable for everyone - there wouldn't be any need for configuration. But that's not the case and there is configuration.

All is good until you miss the one option you need, in which case you either disable that corresponding cop, or if you want to be better then that - you spend the time to reverse-engineer how Rubocop works to modify it to include your use case. And I did exactly that, I wrote tests, documentation, and followed your particular style and recommendations as close as I can.

Saying that this style doesn't makes sense for RuboCop as a whole, kind of violates the spirit of Rubocop, doesn't it? It adds too much personal opinion, in a project where personal opinion should be no more than the default, albeit configurable option.

**Comment by dvandersluis** [comment]
@itay-grudev I'm not suggesting that there should be a single allowed style -- in fact this cop is already configurable as you have seen. That being said, we cannot guarantee that every single style that a company uses makes sense for the RuboCop userbase as a whole. I'm also not saying that this _doesn't_ fit, I'm interested what others on the team think.

We often recommend niche rules be implemented as custom cops (in fact, we have our own custom cops at my workplace that don't necessarily make sense for the tool as a whole). It is simple to create your own cop, and the effort you have put in for this PR would not go to waste, because it could be repurposed as such.

**Comment by bbatsov** [comment]
Indeed. We can't possibly support any style that exists (that would mean an epic overhead in the codebase), that's why we focused only on those that are common within the community. I'm leaning towards rejecting the proposed addition.

**Comment by itay-grudev** [comment]
@bbatsov @dvandersluis We have compact for the following:

`Layout/SpaceInsideArrayLiteralBrackets `
```ruby
array = [ a, [ b, c ]]
array = [[ a ],
  [ b, c ]]
  ```
  
`Layout/SpaceInsideHashLiteralBraces`
```ruby
h = { a: { b: 2 }}
foo = {{ a: 1 } => { b: { c: 2 }}}
```

The `Layout/SpaceInsideParens` `compact` version just adds consistency.

**Comment by bbatsov** [comment]
Hmm, I don't recall when those were added. I'll have to dig up the relevant tickets. 

> The Layout/SpaceInsideParens compact version just adds consistency.

I guess now I'm more receptive to accepting this, but I first have to recall what exactly was I thinking back in the day for those other cops. Perhaps back then I still wasn't as afraid of complexity. :D

**Comment by bbatsov** [comment]
Looking at [this ticket](https://github.com/rubocop/rubocop/issues/4811) things started from the `SpaceInsideHashLiteralBraces` cop and were just copied everywhere else. More digging is needed!

**Comment by itay-grudev** [comment]
@bbatsov In any way - while not as popular, this pattern actually improves readability. Here is an excerpt from our internal guidelines:

> Numerous neuroscience researchers showed that the way the human brain processes text is by looking at a words as a whole rather than individual letters. By improving contrast between different words we help our brain read faster. For example, reading the following sentence is significantly harder when written like this:
> 
> Whitespace(is)oneofthe$most%important#tools@in!our+arsenal;to/increase>code>readability, and it all comes to increasing contrast between different words in a sentence. For example, code highlighting was invented just for that purpose. And while highlighting is usually sufficient to add contrast, adding whitespace correctly further improves it. By employing both syntax highlighting and correct whitespace usage we can help our brains to read code diagonally much faster.
> 
> So the real question then is - where do we add whitespace? What do we add contrast to? What is important?
> 
> Like every language, programming languages have grammar. Grammar is important for compilers or interpreters to understand our code correctly, but it's not that important for humans. We can still understand a sentence with a grammatical error, while a computer won't. For this reason, grammar in a programming language is more important for computers than for humans for reading and understanding a piece of code, hence why grammar is considered low priority when it comes to contrast.
> 
> Instead we should put contrast to part of our code that have meaning to us - like method and variable names, expression operators, values, etc. and not brackets, flow control operators, etc.
> 
> Across the app (_Front-end_ and _Back-end_), we believe spaces are key to code readability. Think of your code as important words or tokens with some grammar in-between. The main rule is that grammar is not as important as content and we add spaces to highlight words, not grammatical structures. Consider the following:
> 

> 
> ```ruby
> #   ╔ language grammar (surrounded by spaces, but note - the whole expression including the opening bracket)
> #   ║       ╔ variable name (surrounded by spaces)
> #   ║       ║    ╔ expression operator (surrounded by spaces as it is important for understanding the context of the expression)
> #   ║       ║    ║       ╔ string value (surrounded by spaces)
> #   ▽       ▽    ▽       ▽
>     if( variable == "expression" ){
> #     △                           △
> #     ╚ a bracket by itself is not a word and adds very little meaning. It's just part of the language's grammar, it is expected there - so no spaces.
> ```

**Comment by flanger001** [comment]
> RubCop is a collection of rules, configurable to suit different projects and different teams with different views.

Exactly, and it is possible to write custom cops for your team for exactly this reason. People submit custom cops here all the time. Sometimes they get added, sometimes they don't.

**Comment by bbatsov** [comment]
I'd suggest taking a look here https://docs.rubocop.org/rubocop/1.19/index.html#philosophy

> The early feedback we got lead us to adopt of philosophy of (extreme) configurability and flexibility, and trying to account for every common style of programming in Ruby.

"common" is the key word here. Anyways, I'll admit I've not always thought enough about some of the things I've accepted, plus over time, as the project grew I became significantly more concerned about the overhead of maintaining some niche styles. 

Again, I'm not rejecting this particular proposal (I'll likely merge it if this has existed in similar cops), I just want everyone to know what we're aiming for in general.

**Comment by koic** [comment]
This is a perspective on the keyword "common". IMHO, I've never encountered this style in well-known programming books and OSS of Ruby. (I may have seen it in Perl...?, but I don't remember it well.) Of course this is my personal perspective. I know that each programmer encounters different styles.

OTOH, I agree with @dvandersluis as a maintainer. I also have several niche custom cops at my workplace.

**Comment by bbatsov** [comment]
@koic I think we're all in agreement here, but as @itay-grudev pointed out somehow this style got added to two other existing cops (`Layout/SpaceInsideArrayLiteralBrackets` and `Layout/SpaceInsideHashLiteralBraces`). For the sake of consistency we either have to add it here as well or remove/deprecate it from the other cops. I have no idea why we felt those braces needed special handling in literals, but not in general, back in the day.

**Comment by koic** [comment]
@bbatsov Thank you for your explanation. My understanding has caught up. But this is certainly a difficult issue to choose. I haven't come up with a better answer yet.
This is a maintenance perspective. `compact` style may be relatively accepted if the alternative option doesn't increase in the future. However, the tendency cannot be confirmed.

**Comment by bbatsov** [comment]
True. At some point we'll have to evaluate all cops/supported styles and remove uncommon stuff to improve the long-term maintainability of the project.

**Comment by dvandersluis** [comment]
I just want to point out that the style on the other cops are symmetrical (applies to both left and right parens) whereas this one is not (only applying to right parens). If we add this I still think it should also be symmetrical to match?

**Comment by itay-grudev** [comment]
@dvandersluis Well, if you agree to merge it then I'll spend the extra time to make those changes and test them, so it's consistent with the other two cops.

**Comment by dvandersluis** [comment]
Up to @bbatsov if we want to continue with this or deprecate the style from the other cops I guess.

**Comment by bbatsov** [comment]
I've decided to continue with adding `compact` here, as the less disruptive change for the users. And to be more careful about the styles we accept going forward as well. :D

**Comment by dvandersluis** [comment]
Alright @itay-grudev please update to make this symmetrical then!

**Comment by itay-grudev** [comment]
@dvandersluis I will finish it later this week and I will ping you when it's ready. Thanks a lot guys!

**Comment by itay-grudev** [comment]
@dvandersluis Implemented left side compaction. I also moved the correction code inside each respective style processor because I started tripping `Metrics/MethodLength` and needed to shorten the code.

File: lib/rubocop/cop/layout/space_inside_parens.rb
Comment (bbatsov) [inline lib/rubocop/cop/layout/space_inside_parens.rb:38]: I think you should remove "right" here. It reads better to me without it.

**Review by bbatsov — COMMENTED** [review]
Looks good, apart from my small remark.

File: lib/rubocop/cop/layout/space_inside_parens.rb
Comment (itay-grudev) [reply lib/rubocop/cop/layout/space_inside_parens.rb:38]: On it.

Comment (itay-grudev) [reply lib/rubocop/cop/layout/space_inside_parens.rb]: Done.

**Review by bbatsov — APPROVED** [review]

**Pull request merged**
